tools/asset-sync/assets.ps1

```diff
@@ -0,0 +1,552 @@
+is present A "root" directory is one where a assets.json is present OR where we are as far up the file tree as 
+we can possibly ascend. 
+
+.PARAMETER TargetPath
+A targeted directory. This MUST be a directory, not a file path.
+#>
+Function EvaluateDirectory {
+    param (
+        [Parameter(Mandatory = $true)]
+        [string] $TargetPath
+    )
+    if (!(Test-Path $TargetPath)){
+        return @(
+            $false, $false
+        )
+    }
+
+    # can't handle files here
+    if (Test-Path $TargetPath -PathType Leaf) {
+        throw "Evaluated a file `"$TargetPath`" as a directory. Exiting."
+    }
+
+    return @(
+        (Test-Path -Path (Join-Path $TargetPath "assets.json")),
+        (Test-Path -Path (Join-Path $TargetPath ".git"))
+    )
+}
+
+<#
+.SYNOPSIS
+From a start directory, stops when it finds the root of a git repository (or root of disk).
+
+.PARAMETER StartPath
+The starting directory.
+#>
+Function AscendToRepoRoot {
+    param(
+        [Parameter(Mandatory = $false)]
+        [string] $StartPath
+    )
+    $pathForManipulation = $StartPath
+    if(Test-Path $StartPath -PathType Leaf){
+        $pathForManipulation = Split-Path $pathForManipulation
+    }
+
+    $foundConfig, $reachedRoot = EvaluateDirectory -TargetPath $pathForManipulation
+
+    while (-not $reachedRoot){
+        $pathForManipulation, $remainder = Split-Path $pathForManipulation
+
+        $foundConfig, $reachedRoot = EvaluateDirectory -TargetPath $pathForManipulation
+    }
+
+    return $pathForManipulation
+}
+
+
+<#
+.SYNOPSIS
+Traverses up from a provided target directory to find a assets JSON, parses it, and returns the location and JSON content.
+
+.DESCRIPTION
+Traverses upwards until it hits either a `.git` folder or a `assets.json` file. Throws an exception if it can't find a assets.json before it hits root.
+
+.PARAMETER TargetPath
+Optional argument specifying the directory to start traversing up from. If not provided, current working directory will be used.
+#>
+Function ResolveAssetsJson {
+    param (
+        [Parameter(Mandatory = $false)]
+        [string] $TargetPath
+    )
+    $pathForManipulation = $TargetPath
+    $foundConfig = $false
+    $reachedRoot = $false
+
+    if(-not $TargetPath){
+        $pathForManipulation = Get-Location
+    }
+    
+    $foundConfig, $reachedRoot = EvaluateDirectory -TargetPath $pathForManipulation
+
+    while (-not $foundConfig -and -not $reachedRoot){
+        $pathForManipulation, $remainder = Split-Path $pathForManipulation
+
+        $foundConfig, $reachedRoot = EvaluateDirectory -TargetPath $pathForManipulation
+    }
+
+    if ($foundConfig){
+        $discoveredPath = Join-Path $pathForManipulation "assets.json"
+    }
+    else {
+        throw "Unable to locate assets.json"
+    }
+
+    # path to assets Json
+    $config = (Get-Content -Raw -Path $discoveredPath | ConvertFrom-Json)
```

Comment:
We probably want to add some validation to make sure this file contains the structure/properties that we expect to find.

Reply:
Good point. 👍